{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#note requires xlsxwriter\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import SeqIO as seqio\n",
    "import math\n",
    "import pandas.io.formats.excel\n",
    "import numpy as np\n",
    "import requests\n",
    "import os\n",
    "from tqdm.notebook import tqdm_notebook\n",
    "from bisect import bisect_left\n",
    "from mpire import WorkerPool\n",
    "from scipy.stats import hypergeom\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "import statistics as stat\n",
    "from scipy.stats import norm\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "pd.io.formats.excel.ExcelFormatter.header_style = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create directory called \"out\" if it doesn't exist\n",
    "if not os.path.exists(\"out\"):\n",
    "    os.makedirs(\"out\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_file(url : str,\n",
    "                filename : str):\n",
    "    headers = {\n",
    "        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36'\n",
    "    }\n",
    "    with requests.get(url, headers=headers, stream=True) as response:\n",
    "        response.raise_for_status()\n",
    "        with open(filename, 'wb') as file:\n",
    "            for chunk in response.iter_content(chunk_size=8192): \n",
    "                if chunk: \n",
    "                    file.write(chunk)\n",
    "\n",
    "def read_matrix_df(scoring_matrices) :\n",
    "    kin_mat_excel = pd.ExcelFile(scoring_matrices, engine=\"openpyxl\")\n",
    "    kinase_names = kin_mat_excel.sheet_names\n",
    "    kin_mat_dfs = {k:kin_mat_excel.parse(k) for k in kinase_names}\n",
    "\n",
    "    any(d.set_index('AA', inplace=True) for d in kin_mat_dfs.values())\n",
    "    kin_mat_dicts = {k:d.to_dict() for k,d in kin_mat_dfs.items()}\n",
    "    kin_mat_dicts = {k:{int(c):aa for c,aa in cl.items()} for k,cl in kin_mat_dicts.items()}\n",
    "    kin_mat_col = {k:list(d.keys()) for k,d in kin_mat_dicts.items()}\n",
    "    kin_mat_col = {k:[int(c) for c in cl] for k,cl in kin_mat_col.items()}\n",
    "    any(c.sort() for c in kin_mat_col.values())\n",
    "\n",
    "    return (kinase_names, kin_mat_dicts, kin_mat_col)\n",
    "\n",
    "def score_with_affinity(sequence, col_map, matrix, use_log=False, ignore_mid=True, densitometry=None) :\n",
    "    mid_res = sequence[5]\n",
    "    no_mid_seq = sequence[:5] + sequence[6:]\n",
    "    \n",
    "    scores = [matrix[col_map[i]][aa] for i,aa in enumerate(no_mid_seq) if (aa != '_')]\n",
    "    if not ignore_mid and densitometry != None:\n",
    "        S_S = sum([densitometry[c]['S'] for c in col_map])\n",
    "        S_T = sum([densitometry[c]['T'] for c in col_map])\n",
    "\n",
    "        S_ctrl = 0.75 * S_S - 0.25 * S_T\n",
    "        T_ctrl = 0.75 * S_T - 0.25 * S_S\n",
    "\n",
    "        S_0 = S_ctrl / (max(S_ctrl, T_ctrl))\n",
    "        T_0 = T_ctrl / (max(S_ctrl, T_ctrl))\n",
    "        \n",
    "        ST_favorability = {'S' : S_0, 'T' : T_0}\n",
    "        if mid_res not in ST_favorability:\n",
    "            print('non-S/T mid: %s' %sequence)\n",
    "            zero_pos = 1\n",
    "        else: \n",
    "            zero_pos = ST_favorability[mid_res]\n",
    "    else:\n",
    "        zero_pos = 1\n",
    "\n",
    "    if use_log :\n",
    "        scores = [math.log2(s) for s in scores]\n",
    "        zero_pos = math.log2(zero_pos)\n",
    "        score = sum(scores) + zero_pos\n",
    "    else :\n",
    "        score = math.prod(scores) * zero_pos\n",
    "\n",
    "    return score\n",
    "\n",
    "def get_percentile(score, background) :\n",
    "    i = bisect_left(background, score)\n",
    "    percentile = i/len(background) * 100\n",
    "    return percentile"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# rearrange PSSM files for easier processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rearrange_matrices (matrices_file : str = './out/johnson_matrices.xlsx',\n",
    "                        sheet_name : str = 'ser_thr_all_norm_scaled_matrice', \n",
    "                    output_file : str = './out/ST-Kinases.xlsx'):\n",
    "    ae_df = pd.read_excel(matrices_file, engine='openpyxl', sheet_name=sheet_name)\n",
    "    #rename first column to Kinase\n",
    "    ae_df.rename(columns={ae_df.columns[0]: 'Kinase'}, inplace=True)\n",
    "    ae_df.set_index('Kinase', inplace=True)\n",
    "\n",
    "    pos = ['-5', '-4', '-3', '-2', '-1', '1', '2', '3', '4']\n",
    "    res = ['P','G','A','C','S','T','V','I','L','M','F','Y','W','H','K','R','Q','N','D','E','s','t','y']\n",
    "\n",
    "    kinase_matrices = {}\n",
    "    for k,row in ae_df.iterrows() :\n",
    "        probs = row.to_numpy()\n",
    "        prob_matrix = np.reshape(probs, (9,23))\n",
    "        prob_matrix_t = prob_matrix.transpose()\n",
    "        kdf = pd.DataFrame(prob_matrix_t, columns=pos, index=res)\n",
    "        kdf.index.name = 'AA'\n",
    "        kinase_matrices[k] = kdf\n",
    "\n",
    "    with pd.ExcelWriter(output_file, engine='xlsxwriter') as writer:\n",
    "        any(df.to_excel(writer, sheet_name=k) for k, df in kinase_matrices.items())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "johnson_matrices_url = 'https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-022-05575-3/MediaObjects/41586_2022_5575_MOESM4_ESM.xlsx'\n",
    "johnson_matrices_original_file = './out/johnson_matrices.xlsx'\n",
    "\n",
    "if not os.path.exists(johnson_matrices_original_file):\n",
    "    download_file(johnson_matrices_url, johnson_matrices_original_file)\n",
    "\n",
    "johnson_matrices_file = './out/ST-Kinases.xlsx'\n",
    "rearrange_matrices(output_file=johnson_matrices_file)\n",
    "\n",
    "densitometry_file = './out/ST-Kinases_densitometry.xlsx'\n",
    "rearrange_matrices(sheet_name = 'ser_thr_all_raw_matrices', output_file=densitometry_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create background file using Ochoa et al data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_ochoa_background(orginal_background_file : str = './out/ochoa_background.xlsx',\n",
    "                        sheet_name : str = 'Supplementary Table 3',\n",
    "                        matrices_file : str = './out/ST-Kinases.xlsx',\n",
    "                        densitometry_file : str = './out/ST-Kinases_densitometry.xlsx',\n",
    "                        output_file : str = './out/johnson_ochoa_background_wfav_log.tsv'):\n",
    "    \n",
    "    #read the original background file\n",
    "    #original file is corrupted and needs to be read with calamine instead of openpyxl\n",
    "    ochoa_df = pd.read_excel(orginal_background_file, engine='calamine', sheet_name=sheet_name)\n",
    "    ochoa_df = ochoa_df[['Database Uniprot Accession', 'Uniprot Primary Accession', 'Protein', 'Description', 'Phosphosite','SITE_+/-7_AA']]\n",
    "    ochoa_df['SITE_+4/-5_AA'] = ochoa_df['SITE_+/-7_AA'].apply(lambda x: x[2:-3].upper())\n",
    "    kinase_names, kin_mat_dicts, kin_mat_col = read_matrix_df(matrices_file)\n",
    "    _, den_dat_dicts, _ = read_matrix_df(densitometry_file)\n",
    "    input_df_scores = pd.concat([pd.Series(ochoa_df['SITE_+4/-5_AA'].apply(lambda x: score_with_affinity(x,\n",
    "        kin_mat_col[k], kin_mat_dicts[k], use_log=False, ignore_mid=False, densitometry = den_dat_dicts[k])),\n",
    "        index=ochoa_df.index, name=k+'_score') for k in tqdm_notebook(kinase_names)], axis=1)\n",
    "    input_df_log2 = pd.concat([pd.Series(ochoa_df['SITE_+4/-5_AA'].apply(lambda x: score_with_affinity(x,\n",
    "                kin_mat_col[k], kin_mat_dicts[k], use_log=True, ignore_mid=False, densitometry = den_dat_dicts[k])),\n",
    "                index=ochoa_df.index, name=k+'_log2') for k in tqdm_notebook(kinase_names)], axis=1)\n",
    "    \n",
    "    ochoa_df = pd.concat([ochoa_df, input_df_scores, input_df_log2], axis=1)\n",
    "\n",
    "    ochoa_df.to_csv(output_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# original url is corrupted\n",
    "ochoa_background_url = 'https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-022-05575-3/MediaObjects/41586_2022_5575_MOESM5_ESM.xlsx'\n",
    "\n",
    "\n",
    "ochoa_background_original_file = './out/ochoa_background.xlsx'\n",
    "ochoa_background_file = './out/johnson_ochoa_background_wfav_log.tsv'\n",
    "\n",
    "\n",
    "if not os.path.exists(ochoa_background_original_file):\n",
    "    download_file(ochoa_background_url, ochoa_background_original_file)\n",
    "\n",
    "if not os.path.exists(ochoa_background_file) :\n",
    "    #takes a while to run (5+ minutes)\n",
    "    build_ochoa_background(orginal_background_file=ochoa_background_original_file, matrices_file = johnson_matrices_file, densitometry_file = densitometry_file, output_file=ochoa_background_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Score and peptides of optosos dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def score_data_file(data_file : str = './input/siteQuant_optosos10m_fracnofrac_combined_no_scan_scrambling_motifsuniquepaste.xlsx',\n",
    "                    sheet_name : str = 'Sheet1',\n",
    "                    matrices_file : str = './out/ST-Kinases.xlsx',\n",
    "                    densitometry_file : str = './out/ST-Kinases_densitometry.xlsx',\n",
    "                    background_file : str = './out/johnson_ochoa_background_wfav_log.tsv',\n",
    "                    output_file : str = './out/siteQuant_optosos10m_fracnofrac_combined_no_scan_scrambling_motifsuniquepaste-all_johnson_ochoa_wfav_log.tsv',\n",
    "                    skip_log2: bool = False,\n",
    "                    ranks_only : bool = False,\n",
    "                    sequence_as_is : bool = False) :\n",
    "        \n",
    "        phospho_df = pd.read_excel(data_file, engine='openpyxl', sheet_name=sheet_name)\n",
    "        if not sequence_as_is:\n",
    "                phospho_df['Motif'] = phospho_df['Motif'].apply(lambda x: x.replace('x','_'))\n",
    "                phospho_df['SITE_+4/-5_AA'] = phospho_df['Motif'].apply(lambda x: x[1:-2])\n",
    "        STonly = phospho_df['SITE_+4/-5_AA'].apply(lambda x: x[5] == 'S' or x[5] == 'T')\n",
    "        phospho_df = phospho_df[STonly].copy()\n",
    "        \n",
    "        kinase_names, kin_mat_dicts, kin_mat_col = read_matrix_df(matrices_file)\n",
    "        _, den_dat_dicts, _ = read_matrix_df(densitometry_file)\n",
    "        phospho_df_scores = pd.concat([pd.Series(phospho_df['SITE_+4/-5_AA'].apply(lambda x: score_with_affinity(x,\n",
    "                kin_mat_col[k], kin_mat_dicts[k], use_log=False, ignore_mid=False, densitometry = den_dat_dicts[k])),\n",
    "                index=phospho_df.index, name=k+'_score') for k in tqdm_notebook(kinase_names)], axis=1)\n",
    "        \n",
    "        if not skip_log2:\n",
    "                phospho_df_log2 = pd.concat([pd.Series(phospho_df['SITE_+4/-5_AA'].apply(lambda x: score_with_affinity(x,\n",
    "                        kin_mat_col[k], kin_mat_dicts[k], use_log=True, ignore_mid=False, densitometry = den_dat_dicts[k])),\n",
    "                        index=phospho_df.index, name=k+'_log2') for k in tqdm_notebook(kinase_names)], axis=1)\n",
    "    \n",
    "                phospho_df = pd.concat([phospho_df, phospho_df_scores, phospho_df_log2], axis=1)\n",
    "        else :\n",
    "\n",
    "                phospho_df = pd.concat([phospho_df, phospho_df_scores], axis=1)\n",
    "\n",
    "        background_df = pd.read_csv(background_file, sep='\\t')\n",
    "        background_df = background_df[[k+'_score' for k in kinase_names]].copy()\n",
    "\n",
    "        kin_dists = {k:list(background_df[k+'_score']) for k in kinase_names}\n",
    "        any(d.sort() for d in kin_dists.values())\n",
    "\n",
    "        phospho_df_percentiles = pd.concat([pd.Series(phospho_df[k+'_score'].apply(lambda x: get_percentile(x, kin_dists[k])),\n",
    "                index=phospho_df.index, name=k+'_percentile') for k in tqdm_notebook(kinase_names)], axis=1)\n",
    "        phospho_df = pd.concat([phospho_df, phospho_df_percentiles], axis=1)\n",
    "\n",
    "        if ranks_only:\n",
    "                phospho_df = pd.concat([phospho_df['SITE_+4/-5_AA'], phospho_df_percentiles], axis=1)\n",
    "        phospho_df.to_csv(output_file, index=False, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "optosos_data = './input/siteQuant_optosos10m_fracnofrac_combined_no_scan_scrambling_motifsuniquepaste.xlsx'\n",
    "\n",
    "optosos_scored = './out/siteQuant_optosos10m_fracnofrac_combined_no_scan_scrambling_motifsuniquepaste-all_johnson_ochoa_wfav_log.tsv'\n",
    "\n",
    "if not os.path.exists(optosos_scored):\n",
    "    #takes a while to run (1+ minute)\n",
    "    score_data_file(data_file=optosos_data, matrices_file = johnson_matrices_file, densitometry_file = densitometry_file, background_file = ochoa_background_file, output_file=optosos_scored)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocess for building figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def binarize(percentile_df, percentile=90) :\n",
    "  binarized_df = pd.DataFrame(np.where(percentile_df >= percentile, True, False),\n",
    "  index=percentile_df.index, columns=percentile_df.columns)\n",
    "  \n",
    "  binarized_df['Any'] = binarized_df.any(axis=1)\n",
    "  binarized_df.reset_index(inplace=True, drop=True)\n",
    "  binarized_df = binarized_df[binarized_df['Any']].copy()\n",
    "  binarized_df.drop(['Any'], axis=1, inplace=True)\n",
    "  \n",
    "  \n",
    "  return binarized_df\n",
    "\n",
    "def create_binarize_file(optosos_file : str = './out/siteQuant_optosos10m_fracnofrac_combined_no_scan_scrambling_motifsuniquepaste-all_johnson_ochoa_wfav_log.tsv',\n",
    "                          fly_orthologs_file : str = './input/uniprotIDs_fly_homologs_non_transmembrane_johnson_kinases.xlsx',\n",
    "                          output_file : str = './out/siteQuant_optosos10m_binarize.csv') :\n",
    "    \n",
    "    cycle_without_membrane = fly_orthologs_file\n",
    "    cycle_without_membrane_df = pd.read_excel(cycle_without_membrane, engine='openpyxl')\n",
    "    cycle_without_membrane_df.set_index('Fly Uniprot ID', inplace=True)\n",
    "    kinases = list(cycle_without_membrane_df['Human kinase name'])\n",
    "\n",
    "    optosos_10m_df = pd.read_csv(optosos_file, sep='\\t')\n",
    "    optosos_10m_df = optosos_10m_df[optosos_10m_df['Max Score'] >= 13].copy()\n",
    "    \n",
    "    experiment_df = optosos_10m_df.filter(regex='_percentile')\n",
    "    experiment_df.columns = [c.split('_percentile')[0] for c in experiment_df.columns]\n",
    "    experiment_df = experiment_df[kinases].copy()\n",
    "\n",
    "    exp_binarize_df = binarize(experiment_df)\n",
    "\n",
    "    #change all columns from bool to int\n",
    "    exp_binarize_num_df = exp_binarize_df.astype(int)\n",
    "    \n",
    "    exp_binarize_num_df.to_csv(output_file, index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "fly_orthologs_file : str = './input/uniprotIDs_fly_homologs_non_transmembrane_johnson_kinases.xlsx'\n",
    "binarize_file = './out/siteQuant_optosos10m_binarize.csv'\n",
    "create_binarize_file(optosos_file=optosos_scored, fly_orthologs_file=fly_orthologs_file, output_file=binarize_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# get background phosphoproteome from expressed genes in early fly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_peptide_seqs(seq, left_offset = 5 , right_offset = 4) :\n",
    "    seq_len = len(seq)\n",
    "    s_or_t_i = [i for i,aa in enumerate(seq) if (aa == 'S') or (aa == 'T')]\n",
    "    s_or_t_i = [i for i in s_or_t_i if (i - left_offset >= 0) and (i+right_offset < seq_len)]\n",
    "    s_or_t_i = [''.join([seq[i-left_offset:i+right_offset+1]]) for i in s_or_t_i]\n",
    "    s_or_t_i = [s for s in s_or_t_i if ('U' not in s) and ('X' not in s)]\n",
    "    return s_or_t_i\n",
    "\n",
    "def build_phosphoproteome(proteome_file : str,\n",
    "        threads : int = 8,\n",
    "        output_file : str = './out/early_fly_phosphoproteome.xlsx ') :\n",
    "    seqrec_dict = seqio.to_dict(seqio.parse(proteome_file, 'fasta'), \n",
    "            key_function= lambda x: x.id.split('|')[1])\n",
    "    seqrec_dict = {k:str(s.seq.upper()) for k,s in seqrec_dict.items()}\n",
    "\n",
    "    seqs = [list(s) for s in seqrec_dict.values()]\n",
    "    seqs = [''.join(s).rstrip() for s in seqs]\n",
    "\n",
    "\n",
    "    with WorkerPool(n_jobs=threads) as executor :\n",
    "        st_seqs = executor.map(get_peptide_seqs, seqs, progress_bar=True)\n",
    "    st_seqs = [s for sq in st_seqs for s in sq]\n",
    "    print(f'Number of 10-mer phosphopeptides : {len(st_seqs)}')\n",
    "    seq_s = pd.Series(st_seqs, name='SITE_+4/-5_AA')\n",
    "    seq_df = pd.concat([seq_s], axis=1)\n",
    "\n",
    "    seq_df.to_excel(output_file, index = False, sheet_name = 'Sheet1', engine = 'xlsxwriter')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "early_fly_proteome_fasta = './input/early_fly_proteome.fasta'\n",
    "early_fly_phosphoproteome = './out/early_fly_phosphoproteome.xlsx'\n",
    "early_fly_phosphoproteome_scored = './out/early_fly_phosphoproteome-all_johnson_ochoa_wfav_ranks.tsv'\n",
    "\n",
    "\n",
    "if not os.path.exists(early_fly_phosphoproteome):\n",
    "    build_phosphoproteome(proteome_file=early_fly_proteome_fasta, output_file=early_fly_phosphoproteome, threads = 1)\n",
    "if not os.path.exists(early_fly_phosphoproteome_scored):\n",
    "    #takes a while to run (15+ minutes)\n",
    "    score_data_file(data_file=early_fly_phosphoproteome, matrices_file = johnson_matrices_file, densitometry_file = densitometry_file, background_file = ochoa_background_file, output_file=early_fly_phosphoproteome_scored, skip_log2=True, ranks_only=True, sequence_as_is=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Do enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kinase_count(percentile_df, percentile=90, normalize=True, binarized=False) :\n",
    "  if not binarized :\n",
    "    binarized_df = binarize(percentile_df, percentile)\n",
    "  else :\n",
    "    binarized_df = percentile_df.copy()\n",
    "  total_peptides = len(binarized_df.index)\n",
    "\n",
    "  binarized_df = binarized_df.transpose()\n",
    "\n",
    "  counts = binarized_df.sum(axis=1)\n",
    "  if(normalize) :\n",
    "    counts = np.divide(counts, total_peptides)\n",
    "  counts_s = pd.Series(counts, index=binarized_df.index, name='c%d' % (percentile))\n",
    "  return counts_s\n",
    "\n",
    "def do_enrichment_analysis(experiment_df, background_df, fdr=0.1) :\n",
    "  exp_binarize_df = binarize(experiment_df)\n",
    "  num_experiment_peptides = len(exp_binarize_df.index)\n",
    "\n",
    "  bak_binarize_df = binarize(background_df)\n",
    "  num_background_peptides = len(bak_binarize_df.index)\n",
    "\n",
    "  experiment_count = kinase_count(exp_binarize_df, normalize=False, binarized=True)\n",
    "  experiment_count.rename('experiment_count',inplace=True)\n",
    "\n",
    "  background_count = kinase_count(bak_binarize_df, normalize=False, binarized=True)\n",
    "  background_count.rename('background_count',inplace=True)\n",
    "\n",
    "  experiment_enrich = kinase_count(exp_binarize_df, binarized=True)\n",
    "  experiment_enrich.rename('experiment_ratio',inplace=True)\n",
    "\n",
    "  background_enrich = kinase_count(bak_binarize_df, binarized=True)\n",
    "  background_enrich.rename('background_ratio',inplace=True)\n",
    "\n",
    "  count_df = pd.concat([background_count, experiment_count, background_enrich, experiment_enrich], axis=1)\n",
    "  count_df['enrichment'] = np.divide(count_df['experiment_ratio'], count_df['background_ratio'])\n",
    "\n",
    "  count_df['cdf'] = hypergeom.cdf(k = count_df['experiment_count'],\n",
    "                  M = num_background_peptides,\n",
    "                  n = count_df['background_count'],\n",
    "                  N = num_experiment_peptides)\n",
    "\n",
    "  count_df['pmf'] = hypergeom.pmf(k = count_df['experiment_count'],\n",
    "                  M = num_background_peptides,\n",
    "                  n = count_df['background_count'],\n",
    "                  N = num_experiment_peptides)\n",
    "  \n",
    "  count_df['p-value'] = count_df.apply(lambda x: 1 - x['cdf'] + x['pmf'] if x['enrichment'] >= 1.0 else x['cdf'], axis=1)\n",
    "\n",
    "  count_df['fdr'] = fdrcorrection(count_df['p-value'], alpha=fdr)[1]\n",
    "\n",
    "  count_df.drop(['cdf', 'pmf'], axis=1, inplace=True)\n",
    "\n",
    "  count_df.sort_values(by=['enrichment'], ascending=False, inplace=True)\n",
    "\n",
    "\n",
    "  return (num_experiment_peptides, num_background_peptides, count_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_dfs = {}\n",
    "peptide_counts = []\n",
    "fdr_cutoffs = {'topright_vs_all' : 20, 'all_vs_proteome' : 0.10}\n",
    "\n",
    "cycle_without_membrane = fly_orthologs_file\n",
    "cycle_without_membrane_df = pd.read_excel(cycle_without_membrane, engine='openpyxl')\n",
    "cycle_without_membrane_df.set_index('Fly Uniprot ID', inplace=True)\n",
    "kinases = list(cycle_without_membrane_df['Human kinase name'])\n",
    "\n",
    "name = 'topright_vs_all'\n",
    "optosos_10m_df = pd.read_csv(optosos_scored, sep='\\t')\n",
    "optosos_10m_df = optosos_10m_df[optosos_10m_df['Max Score'] >= 13].copy()\n",
    "optosos_10m_top_right_df = optosos_10m_df[(optosos_10m_df['neglogadjp'] >= 2) & (optosos_10m_df['log2FC'] >= 0)].copy()\n",
    "\n",
    "optosos_10m_df  = optosos_10m_df.filter(regex='_percentile')\n",
    "optosos_10m_df.columns = [c.split('_percentile')[0] for c in optosos_10m_df .columns]\n",
    "optosos_10m_df = optosos_10m_df[kinases].copy()\n",
    "\n",
    "optosos_10m_top_right_df  = optosos_10m_top_right_df.filter(regex='_percentile')\n",
    "optosos_10m_top_right_df.columns = [c.split('_percentile')[0] for c in optosos_10m_top_right_df.columns]\n",
    "optosos_10m_top_right_df = optosos_10m_top_right_df[kinases].copy()\n",
    "\n",
    "\n",
    "(num_optosos_10m_top_right_peptides, num_optosos_10m_peptides, optosos_10m_topright_vs_all_df) = do_enrichment_analysis(optosos_10m_top_right_df, optosos_10m_df, fdr=fdr_cutoffs[name])\n",
    "\n",
    "del optosos_10m_top_right_df\n",
    "\n",
    "#optosos_10m_topright_vs_all_df.index.name = 'kinase'\n",
    "#enrichment_dfs[name] = optosos_10m_topright_vs_all_df\n",
    "#peptide_counts.append((name, num_optosos_10m_top_right_peptides, num_optosos_10m_peptides, fdr_cutoffs[name]))\n",
    "\n",
    "name = 'all_vs_proteome'\n",
    "\n",
    "early_fly_phosphoproteome_df = pd.read_csv(early_fly_phosphoproteome_scored, sep='\\t')\n",
    "early_fly_phosphoproteome_df.drop(list(early_fly_phosphoproteome_df.filter(regex='_score|_log2|SITE_')), axis=1, inplace=True)\n",
    "early_fly_phosphoproteome_df.columns = [c.split('_percentile')[0] for c in early_fly_phosphoproteome_df.columns]\n",
    "early_fly_phosphoproteome_df = early_fly_phosphoproteome_df[kinases].copy()\n",
    "\n",
    "(num_optosos_10m_peptides, num_early_fly_proteome_peptides, optosos_10m_all_vs_proteome_df) = do_enrichment_analysis(optosos_10m_df, early_fly_phosphoproteome_df, fdr=fdr_cutoffs[name])\n",
    "\n",
    "del early_fly_phosphoproteome_df\n",
    "del optosos_10m_df\n",
    "\n",
    "optosos_10m_all_vs_proteome_df.index.name = 'kinase'\n",
    "enrichment_dfs[name] = optosos_10m_all_vs_proteome_df\n",
    "peptide_counts.append((name, num_optosos_10m_peptides, num_early_fly_proteome_peptides, fdr_cutoffs[name]))\n",
    "\n",
    "# Add summary sheet\n",
    "summary_counts_df = pd.DataFrame(peptide_counts, columns=['name', 'num experiment peptides', 'num background peptides', 'fdr threshold'])\n",
    "summary_counts_df.index.name = 'experiment number'\n",
    "sheets_dict = {'summary': summary_counts_df}\n",
    "sheets_dict.update(enrichment_dfs)\n",
    "\n",
    "#combine sheets in Excel file\n",
    "excel_output = './out/enrichment_variations.xlsx'\n",
    "with pd.ExcelWriter(excel_output, engine='xlsxwriter') as writer:\n",
    "  any(df.to_excel(writer, sheet_name=k, index=True) for k, df in sheets_dict.items())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Z-score analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#experiment file\n",
    "optosos_10m_df = pd.read_csv(optosos_scored, sep='\\t')\n",
    "optosos_10m_df = optosos_10m_df[optosos_10m_df['Max Score'] >= 13].copy()\n",
    "\n",
    "#56 fly homologs cytoplasmic kinases\n",
    "cycle_without_membrane = fly_orthologs_file\n",
    "cycle_without_membrane_df = pd.read_excel(cycle_without_membrane, engine='openpyxl')\n",
    "cycle_without_membrane_df.set_index('Fly Uniprot ID', inplace=True)\n",
    "kinases = list(cycle_without_membrane_df['Human kinase name'])\n",
    "\n",
    "experiment_df = optosos_10m_df.filter(regex='log2FC|_percentile')\n",
    "\n",
    "experiment_df.columns = [c.split('_percentile')[0] if '_percentile' in c else c for c in experiment_df.columns]\n",
    "\n",
    "select_columns = ['log2FC']\n",
    "select_columns.extend(kinases)\n",
    "experiment_df = experiment_df[select_columns].copy()\n",
    "\n",
    "pop_mean = stat.mean(experiment_df['log2FC'])\n",
    "pop_stdv = stat.stdev(experiment_df['log2FC'])\n",
    "\n",
    "kin_stats = []\n",
    "for k in kinases :\n",
    "    k_sub = experiment_df[(experiment_df[k] >= 90)]['log2FC'].copy()\n",
    "    k_mean = stat.mean(k_sub)\n",
    "    k_stdv = stat.stdev(k_sub)\n",
    "    k_n = len(k_sub.index)\n",
    "    z = (k_mean - pop_mean) / (pop_stdv / math.sqrt(k_n))\n",
    "    if z > 0 :\n",
    "        p = 1 - norm.cdf(z) + norm.pdf(z)\n",
    "    else :\n",
    "        p = norm.cdf(z)\n",
    "    #print('%s\\t%1.3f\\t%0.3e' %(k,z,p))\n",
    "    kin_stats.append((k, k_n, k_mean, k_stdv, z, p))\n",
    "\n",
    "pop_stats = [(pop_mean, pop_stdv)]\n",
    "\n",
    "pop_df = pd.DataFrame(pop_stats, columns=['mean', 'stdv'])\n",
    "kin_df = pd.DataFrame(kin_stats, columns=['kinase', 'n', 'mean', 'stdv', 'z', 'p-value'])\n",
    "kin_df['fdr'] = fdrcorrection(kin_df['p-value'])[1]\n",
    "\n",
    "excel_output = './out/kinase_z_scores_flip.xlsx'\n",
    "with pd.ExcelWriter(excel_output, engine='xlsxwriter') as writer:\n",
    "  pop_df.to_excel(writer, sheet_name='population', index=False)\n",
    "  kin_df.to_excel(writer, sheet_name='kinases', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
